lib/ansible/module_utils/common/text/converters.py

```diff
@@ -23,7 +30,33 @@
                                       'surrogate_then_replace'))
 
 
-def to_bytes(obj, encoding='utf-8', errors=None, nonstring='simplerepr'):
+@_t.overload
+def to_bytes(
+    obj: _T,
+    encoding: str = 'utf-8',
+    errors: str | None = None,
+    *,
+    nonstring: _NonStringPassthru,
+) -> _T | bytes: ...
```

Comment:
For the `passthru` cases, we should be able to split the `_T | bytes` into two overloads, something like this:

```python
_TSB = _t.TypeVar('_TSB', str, bytes)

@_t.overload
def to_text(
    obj: _TSB,
    encoding: str = 'utf-8',
    errors: str | None = None,
    *,
    nonstring: _NonStringPassthru,
) -> str: ...


@_t.overload
def to_text(
    obj: _T,
    encoding: str = 'utf-8',
    errors: str | None = None,
    *,
    nonstring: _NonStringPassthru,
) -> _T: ...
```

That would eliminate the `_T | bytes` ambiguity in the first overload, leaving it only on the function.

Reply:
Let me stare at it, my brain needs a bit more of a kickstart.

Reply:
Ok, I got it figured out. I think my problem originally when I tried this was also that the order of the overloads matters here.  I've pushed a commit with `_NonStringPassthru` split into 2.